app/src/cose.h

```diff
@@ -60,11 +81,156 @@ namespace scitt::cose
 
   static constexpr const char* SVN_HEADER_PARAM = "svn";
 
+  static std::shared_ptr<ccf::crypto::HashProvider>& get_hash_provider()
+  {
+    static thread_local std::shared_ptr<ccf::crypto::HashProvider>
```

Comment:
I am surprised that this is worth caching, because:

1. HashProvider (OpenSSLHashProvider really) construction is trivial, there is nothing to initialise
2. There are unfortunate lookup costs per call (not per construction) in OpenSSL3, but caching the provider does not help with those and they are mitigated by CCF already (https://github.com/microsoft/CCF/blob/main/src/crypto/openssl/hash.cpp#L59)

Reply:
This was done to be able to do unit tests as they would otherwise fail due to init not called. And init seems to be not exported in a way to be invoked in a test, hence this method :(

Reply:
Ah, I see! Yes, it's not in any exported headers at the moment, but that's something we could fix easily.